### DISCUSSION TITLE
TLS 1.3 offload to PKCS11 Provider

### DISCUSSION THREAD

**Discussion by sahilnxp** [openssl-projects/pkcs11-provider#389] [Q&A]
Had a look at https://github.com/latchset/pkcs11-provider/pull/216
A test to exercise full TLS connection was added in this, looks like we have the Key and Certificates stored in PKCS#11 Token.
If I understood correctly Signing operation with the Key stored in Token will be done by the PKCS#11 implementation.

Apart from this signing, can you please tell me which operations are offloaded to PKCS#11 Provider in Full TLS connection?
Is TLS1.3 Session key derivation is also offloaded to the PKCS#11 implementation?
If yes, Which API is responsible for session key derivation?

@simo5 Can you please look at this?

**Comment by simo5** [comment]
OpenSSL avoids offloading any operation unless it is forced (key only available in provider) or is explicitly instructed to do so via properties.

So by default and in current tests only the Signature operation is offloaded.

That said, PKCS#11 does define a bunch of KDFs for TLS, but pkcs11-provider does not yet wrap all of them.

I am not sure there is a way to tell OpenSSL to offload just the KDFs, you may have to try to offload everything to the token via a default property that applies to all operations. That may or not be desirable.

**Comment by sahilnxp** [comment]
> That said, PKCS#11 does define a bunch of KDFs for TLS, but pkcs11-provider does not yet wrap all of them.

Ok, can you please tell which KDFs pkcs11-provider currently wraps?

> I am not sure there is a way to tell OpenSSL to offload just the KDFs, you may have to try to offload everything to the token via a default property that applies to all operations. That may or not be desirable

Hmm, that needs to be thought, but do you have an example which we can refer for defining a default property that applies to all operations ?

Openssl gives a way to offload the TLS PRF functions to HW crypto, using engines but don't know about the providers.

**Comment by simo5** [comment]
> > That said, PKCS#11 does define a bunch of KDFs for TLS, but pkcs11-provider does not yet wrap all of them.
> 
> Ok, can you please tell which KDFs pkcs11-provider currently wraps?

Only HKDF so far:
https://github.com/latchset/pkcs11-provider/blob/main/src/kdf.h

> > I am not sure there is a way to tell OpenSSL to offload just the KDFs, you may have to try to offload everything to the token via a default property that applies to all operations. That may or not be desirable
> 
> Hmm, that needs to be thought, but do you have an example which we can refer for defining a default property that applies to all operations ?

you can set default property of `provider=pkcs11` in the confgi file or pass it via -propq on command line utilties.

> Openssl gives a way to offload the TLS PRF functions to HW crypto, using engines but don't know about the providers.

See above.
The kdfs openssl support from providers are:
- hkdf
- hmacdrbg
- kbkdf
- krb5kdf
- pbkdf2
- pkcs12kdf
- pvkkdf
- scrypt
- sshkdf
- sskdf
- tls1_prf
- x492kdf

So far we wired only the first one, it is not hard to do many of others (where there is a pkcs#11 spec that covers them), we just haven't done it yet.

**Reply by kshitizvars** [reply]
Hi @simo5 

We are trying to run openssl test mentioned in **tests/thkdf:**-

 ```
export HKDF_HEX_SECRET=ffeeddccbbaa
export HKDF_HEX_SALT=ffeeddccbbaa
export HKDF_HEX_INFO=ffeeddccbbaa
openssl
pkeyutl -derive -kdf HKDF -kdflen 48
                -pkeyopt md:SHA256
                -pkeyopt mode:EXTRACT_AND_EXPAND
                -pkeyopt hexkey:${HKDF_HEX_SECRET}
                -pkeyopt hexsalt:${HKDF_HEX_SALT}
                -pkeyopt hexinfo:${HKDF_HEX_INFO}
                -out ${TMPPDIR}/hkdf1-out-pkcs11.bin
                -propquery provider=pkcs11'
```

**Error Recieved:-**

```
pkeyutl: Error initializing context
20500984FFFF0000:error:03000096:digital envelope routines:EVP_PKEY_derive_init_ex:operation not supported for this keytype:crypto/evp/exchange.c:377:
```

**Debugging done on our end:-**
Issue seems to be in linking provider's key exchange API's to openssl.

**PKCS#11-provider debug logs:-**
[debug.log](https://github.com/latchset/pkcs11-provider/files/15347501/debug.log)

Can you please guide us and let us know, that from where issue is arising? Is it from **openssl pkcs11-provider** linking or **pkcs11-provider  optee** linking.

**Reply by simo5** [reply]
I can't reproduce locally, what OpenSSL version are you using?
Does your token actually implement CKM_HKDF_DERIVE ?

**Reply by kshitizvars** [reply]
Openssl version is **OpenSSL 3.2.1**.
Yes, CKM_HKDF_DERIVE is missing from our token.
Thanks for the heads up!!

**Reply by kshitizvars** [reply]
Hi @simo5,

We have generated token which implement ECDH derive operation:-

Output of `pkcs11-tool --module $PKCS11_MODULE_PATH --list-objects` which clearly shows that object support derive and signing operation:-
 
```
root@imx8ulpevk:~/using_pkcs11_tool# pkcs11-tool --module $PKCS11_MODULE_PATH --list-objects                                                                                            Using slot 0 with a present token (0x0)
Public Key Object; EC  EC_POINT 256 bits
  EC_POINT:   0441041d67914d3949573917aaaa381a9f7a8d6d4ce9d7f3776e5f3112d32c18fa215ca358a3761b3c37c21356346aa22aa3a8bf5bf3799139cd96595e41a35b550e4f
  EC_PARAMS:  06082a8648ce3d030107 (OID 1.2.840.10045.3.1.7)
  label:      ecc-key-256_2
  ID:         01
  Usage:      verify, derive
  Access:     local

```

**We can run below pkeyutl command successfully and ECDH operations are getting offloaded to optee as well:-**

`openssl pkeyutl -derive -inkey "pkcs11:model=OP-TEE%20TA;manufacturer=Linaro;serial=0000000000000000;token=kshitiz_test;object=ecc-key-256_2;type=private" -peerkey server_pub.pem -out secret_derive.ecdh.bin`

But, if we try to establish TLS1.3 connection using **openssl s_server** and **openssl s_client** using same object:-
**Server side command:-**
`openssl s_server -key  "pkcs11:model=OP-TEE%20TA;manufacturer=Linaro;serial=0000000000000000;token=kshitiz_test;id=%01;object=ecc-key-256_2;type=private?pin-value=1234" -cert req_client.crt -accept 443 -Verify 3`

**Client side command:-**
`openssl s_client -connect 10.232.132.242:443 -tls1_3 -ciphersuites 'TLS_AES_256_GCM_SHA384' -cert server.crt -key server.key  -CAfile ca.crt`

**We do not see ECDH operation getting offloaded to optee. Can you please help in the same?**


Pkcs11-provider debug logs:- [debug.log](https://github.com/latchset/pkcs11-provider/files/15389329/debug.log)

**Reply by simo5** [reply]
try adding -propquery provider=?pkcs11

OpenSSL by default *does not* use tokens if it implements a function in the default provider. So you have to try to force it by setting property queries that cause OpenSSL to prefer your token to the default implementation.

**Reply by kshitizvars** [reply]
Still no luck after adding **-propquery provider=?pkcs11** in the openssl s_server command like below:-

`openssl s_server -key  "pkcs11:model=OP-TEE%20TA;manufacturer=Linaro;serial=0000000000000000;token=kshitiz_test;id=%01;object=ecc-key-256_2;type=private?pin-value=1234" -cert req_client.crt -accept 443 -Verify 3 -propquery provider=?pkcs11`

**Reply by simo5** [reply]
You could try adding them as default properties in openssl.cnf file
Not always -propquery is propagated to every call via cmd line tools

**Reply by kshitizvars** [reply]
This is how our openssl.cnf looks like with pkcs11-provider related changes:-

```
[openssl_init]
providers = provider_sect
 
# List of providers to load
[provider_sect]
default = default_sect
pkcs11 = pkcs11_sect
 
[default_sect]
activate = 1
[pkcs11_sect]
module = /usr/lib/pkcs11.so
pkcs11-module-path = /usr/lib/libckteec.so.0
pkcs11-module-cache-keys = false
pkcs11-module-quirks = no-operation-state
activate = 1
```

Can you please point out where exactly we need to add and what needs to be added?

**Reply by simo5** [reply]
Checkout the "EVP Configuration" section in the openssl.cnf manpage

**Reply by kshitizvars** [reply]
Hi

This is my openssl.cnf after applying EVP configuration:-
```
[openssl_init]
providers = provider_sect
alg_section = algorithm_sect

# List of providers to load
[provider_sect]
default = default_sect
pkcs11 = pkcs11_sect

[default_sect]
activate = 1
[pkcs11_sect]
module = /usr/lib/pkcs11.so
pkcs11-module-path = /usr/lib/libckteec.so.0
pkcs11-module-cache-keys = false
pkcs11-module-quirks = no-operation-state
activate = 1

[algorithm_sect]
default_properties = ?provider=pkcs11
```

**Server side command:-**
`openssl s_server -key  "pkcs11:model=OP-TEE%20TA;manufacturer=Linaro;serial=0000000000000000;token=kshitiz_test;id=%02;object=ecc-key-521;type=private?pin-value=1234" -cert req_client.crt -accept 443 -Verify 3 -named_curve secp521r1 -ciphersuites 'TLS_AES_256_GCM_SHA384' -tls1_3`

**Client side command:-**
`./openssl s_client -connect 10.232.132.242:443 -tls1_3 -ciphersuites 'TLS_AES_256_GCM_SHA384' -cert server.crt -key server.key  -CAfile ../ca.crt`

**Error recieved on server side:-**
```
verify depth is 3, must return a certificate
Using default temp DH parameters
ACCEPT
ERROR
20904DB5FFFF0000:error:40800007:pkcs11:p11prov_common_gen_init:Invalid or improper arguments were provided to the invoked function:/usr/src/debug/pkcs11-provider/0.3/src/keymgmt.c:150:Unsupported selection
20904DB5FFFF0000:error:03000086:digital envelope routines:gen_init:initialization error:/usr/src/debug/openssl/3.2.1/crypto/evp/pmeth_gn.c:52:
20904DB5FFFF0000:error:0A00013A:SSL routines:tls_parse_ctos_key_share:unable to find ecdh parameters:/usr/src/debug/openssl/3.2.1/ssl/statem/extensions_srvr.c:684:
```
Am I doing something wrong here? And one more thing, that if we are able to reach to optee in sign verification operation without applying EVP configuration change, so why this change is required for exchange operation? Any specific reason?

**Reply by simo5** [reply]
Not your fault, please open a feature request
You hit an unimplemented Keymgmt feature we need to add to make this work via a token.

**Reply by simo5** [reply]
I actually opened #395
